source/s3_parallel_input_stream.c

```diff
@@ -0,0 +1,165 @@
+    }
+    return stream;
+}
+
+struct aws_parallel_input_stream *aws_parallel_input_stream_release(struct aws_parallel_input_stream *stream) {
+    if (stream != NULL) {
+        aws_ref_count_release(&stream->ref_count);
+    }
+    return stream;
+}
+
+struct aws_future_bool *aws_parallel_input_stream_read(
+    struct aws_parallel_input_stream *stream,
+    uint64_t offset,
+    struct aws_byte_buf *dest) {
+    /* Ensure the buffer has space available */
+    if (dest->len == dest->capacity) {
+        struct aws_future_bool *future = aws_future_bool_new(stream->alloc);
+        aws_future_bool_set_error(future, AWS_ERROR_SHORT_BUFFER);
+        return future;
+    }
+    /* TODO: restore the buffer on failure. */
+    struct aws_future_bool *future = stream->vtable->read(stream, offset, dest);
+    return future;
+}
+
+struct aws_parallel_input_stream_from_file_impl {
+    struct aws_parallel_input_stream base;
+
+    struct aws_string *file_path;
+    uint64_t last_modified_time;
+};
+
+static void s_para_from_file_destroy(struct aws_parallel_input_stream *stream) {
+    struct aws_parallel_input_stream_from_file_impl *impl =
+        AWS_CONTAINER_OF(stream, struct aws_parallel_input_stream_from_file_impl, base);
+
+    aws_string_destroy(impl->file_path);
+
+    aws_mem_release(stream->alloc, impl);
+
+    return;
+}
+
+static int s_get_last_modified_time(const char *file_name, uint64_t *out_time) {
+    struct stat attrib;
+    if (stat(file_name, &attrib)) {
+        return aws_translate_and_raise_io_error(errno);
+    }
+    /* The time stamp in in secs, it's quite a pain to support nanoseconds for different platform, we can support it
+     * when we need to. */
+    *out_time = (uint64_t)attrib.st_mtime;
+    return AWS_OP_SUCCESS;
+}
+
+struct aws_future_bool *s_para_from_file_read(
+    struct aws_parallel_input_stream *stream,
+    uint64_t offset,
+    struct aws_byte_buf *dest) {
+
+    struct aws_future_bool *future = aws_future_bool_new(stream->alloc);
+    struct aws_parallel_input_stream_from_file_impl *impl =
+        AWS_CONTAINER_OF(stream, struct aws_parallel_input_stream_from_file_impl, base);
+    bool success = false;
+    uint64_t last_modified_time = 0;
+    struct aws_input_stream *file_stream = NULL;
+    struct aws_stream_status status = {
+        .is_end_of_stream = false,
+        .is_valid = true,
+    };
+
+    if (s_get_last_modified_time(aws_string_c_str(impl->file_path), &last_modified_time)) {
+        goto done;
+    }
+    /* Check if file modified after we create the input stream */
+    if (last_modified_time != impl->last_modified_time) {
+        aws_raise_error(AWS_ERROR_S3_FILE_MODIFIED);
+        goto done;
+    }
+    file_stream = aws_input_stream_new_from_file(stream->alloc, aws_string_c_str(impl->file_path));
+    if (aws_input_stream_seek(file_stream, offset, AWS_SSB_BEGIN)) {
+        goto done;
+    }
+    /* Keep reading until fill the buffer */
```

Comment:
This code is good. I just wondered if we should be checking for EOF before doing the read(), but googled it and [apparently not](https://stackoverflow.com/questions/17263830/fseek-on-a-position-beyond-eof-does-not-trigger-eof-using-feof-how-come)!

Maybe worth adding a note for anyone else who wonders the same?
```suggestion
    /* Keep reading until fill the buffer.
     * Note that we must read() after seek() to determine if we're EOF, the seek alone won't trigger it. */
```